File: guitar-mobile-expo/modules/expo-pitch-detector/android/src/main/java/expo/modules/pitchdetector/PitchDetectorMPM.kt
package expo.modules.pitchdetector

import kotlin.math.PI
import kotlin.math.cos
import kotlin.math.max
import kotlin.math.min
import kotlin.math.sqrt

data class PitchResult(val frequency: Double, val clarity: Double)

class PitchDetectorMPM(
  private val sampleRate: Double,
  private val n: Int,
) {
  private val clarityFloor = 0.7
  private val kCutoff = 0.93
  private val minFreq = 55.0
  private val maxFreq = 1500.0
  private val window: FloatArray = FloatArray(n) { i ->
    (0.5 * (1 - cos(2 * PI * i / (n - 1)))).toFloat()
  }

  fun detect(samples: FloatArray): PitchResult? {
    if (samples.size < n) return null
    val x = FloatArray(n) { samples[it] * window[it] }

    val maxLag = n / 2
    val minLag = max(2, (sampleRate / maxFreq).toInt())
    val maxAllowedLag = min(maxLag, (sampleRate / minFreq).toInt())

    val nsdf = DoubleArray(maxAllowedLag + 1)
    for (tau in 0..maxAllowedLag) {
      var acf = 0.0
      var m = 0.0
      val upper = n - tau
      var i = 0
      while (i < upper) {
        val a = x[i].toDouble()
        val b = x[i + tau].toDouble()
        acf += a * b
        m += a * a + b * b
        i++
      }
      nsdf[tau] = if (m > 0) 2 * acf / m else 0.0
    }

    // Skip the initial positive lobe — NSDF starts at 1.0 and descends through its first
    // zero-crossing. Without this the peak scan latches onto that descent and reports a
    // spurious high-frequency artifact (e.g. sampleRate / minLag).
    var i = minLag
    while (i < maxAllowedLag && nsdf[i] > 0) i++

    val peaks = mutableListOf<Pair<Int, Double>>()
    while (i < maxAllowedLag) {
      if (nsdf[i] > 0) {
        var j = i
        var bestTau = j
        var bestVal = nsdf[j]
        while (j < maxAllowedLag && nsdf[j] > 0) {
          if (nsdf[j] > bestVal) { bestTau = j; bestVal = nsdf[j] }
          j++
        }
        peaks += bestTau to bestVal
        i = j + 1
      } else {
        i++
      }
    }
    val maxVal = peaks.maxOfOrNull { it.second } ?: return null
    val chosen = peaks.firstOrNull { it.second >= kCutoff * maxVal } ?: return null

    val tau = chosen.first
    val y0 = if (tau > 0) nsdf[tau - 1] else nsdf[tau]
    val y1 = nsdf[tau]
    val y2 = if (tau + 1 <= maxAllowedLag) nsdf[tau + 1] else nsdf[tau]
    val denom = y0 - 2 * y1 + y2
    val shift = if (denom == 0.0) 0.0 else 0.5 * (y0 - y2) / denom
    val tauStar = tau + shift
    val freq = sampleRate / tauStar
    val clarity = y1
    if (clarity < clarityFloor) return null
    if (freq < minFreq || freq > maxFreq) return null
    return PitchResult(freq, clarity)
  }

  fun rms(samples: FloatArray): Double {
    var s = 0.0
    for (v in samples) s += v.toDouble() * v
    return if (samples.isNotEmpty()) sqrt(s / samples.size) else 0.0
  }
}

File: guitar-mobile-expo/modules/expo-pitch-detector/ios/PitchDetectorMPM.swift
// modules/expo-pitch-detector/ios/PitchDetectorMPM.swift
import Accelerate
import Foundation

public struct PitchResult {
  public let frequency: Double
  public let clarity: Double
}

public final class PitchDetectorMPM {
  private let sampleRate: Double
  private let n: Int
  private let window: [Float]
  private let clarityFloor: Double = 0.6
  private let kCutoff: Double = 0.93
  private let minFreq: Double = 55
  private let maxFreq: Double = 1500

  public init(sampleRate: Double, windowSize: Int) {
    self.sampleRate = sampleRate
    self.n = windowSize
    self.window = (0..<windowSize).map { i in
      0.5 * (1 - cos(2 * .pi * Float(i) / Float(windowSize - 1)))
    }
  }

  public func detect(_ samples: [Float]) -> PitchResult? {
    guard samples.count >= n else { return nil }

    // Hann-window the signal (Float), then promote to Double for the NSDF math.
    var windowed = [Float](repeating: 0, count: n)
    vDSP_vmul(samples, 1, window, 1, &windowed, 1, vDSP_Length(n))
    var x = [Double](repeating: 0, count: n)
    vDSP_vspdp(windowed, 1, &x, 1, vDSP_Length(n))

    // Prefix sums of squares make the NSDF normalizer an O(1) lookup:
    //   m(tau) = sum_{i<n-tau} x[i]^2 + sum_{i<n-tau} x[i+tau]^2
    //          = prefix[n-tau] + (prefix[n] - prefix[tau])
    var sq = [Double](repeating: 0, count: n)
    vDSP_vsqD(x, 1, &sq, 1, vDSP_Length(n))
    var prefix = [Double](repeating: 0, count: n + 1)
    var running: Double = 0
    for i in 0..<n {
      running += sq[i]
      prefix[i + 1] = running
    }
    let total = prefix[n]

    let maxLag = n / 2
    let minLag = max(2, Int(sampleRate / maxFreq))
    let maxAllowedLag = min(maxLag, Int(sampleRate / minFreq))

    var nsdf = [Double](repeating: 0, count: maxAllowedLag + 1)
    x.withUnsafeBufferPointer { xp in
      let base = xp.baseAddress!
      for tau in 0...maxAllowedLag {
        var acf: Double = 0
        vDSP_dotprD(base, 1, base + tau, 1, &acf, vDSP_Length(n - tau))
        let m = prefix[n - tau] + (total - prefix[tau])
        nsdf[tau] = (m > 0) ? (2 * acf / m) : 0
      }
    }

    // --- peak picking (unchanged from the scalar implementation) ---
    // Skip the initial positive lobe (NSDF starts at 1.0 and descends through its
    // first zero-crossing) so the scan doesn't latch onto that descent.
    var i = minLag
    while i < maxAllowedLag && nsdf[i] > 0 { i += 1 }

    // Key maxima: highest peak in each positively-sloped region between zero-crossings.
    var peaks: [(tau: Int, val: Double)] = []
    while i < maxAllowedLag {
      if nsdf[i] > 0 {
        var j = i
        var best = (tau: j, val: nsdf[j])
        while j < maxAllowedLag && nsdf[j] > 0 {
          if nsdf[j] > best.val { best = (tau: j, val: nsdf[j]) }
          j += 1
        }
        peaks.append(best)
        i = j + 1
      } else {
        i += 1
      }
    }
    guard let maxVal = peaks.map({ $0.val }).max() else { return nil }
    guard let chosen = peaks.first(where: { $0.val >= kCutoff * maxVal }) else { return nil }

    let tau = chosen.tau
    let y0 = (tau > 0) ? nsdf[tau - 1] : nsdf[tau]
    let y1 = nsdf[tau]
    let y2 = (tau + 1 <= maxAllowedLag) ? nsdf[tau + 1] : nsdf[tau]
    let denom = (y0 - 2 * y1 + y2)
    let shift = denom == 0 ? 0 : 0.5 * (y0 - y2) / denom
    let tauStar = Double(tau) + shift

    let freq = sampleRate / tauStar
    let clarity = y1
    if clarity < clarityFloor { return nil }
    if freq < minFreq || freq > maxFreq { return nil }
    return PitchResult(frequency: freq, clarity: clarity)
  }
}
